### ISSUE TITLE
Netplay lobby should organize by ping or region

### ISSUE THREAD

**Issue by silentq** [libretro/RetroArch#4567]
To prevent bad or slow connections it may be easier for the user to have some idea of the geographic region the host is in before connecting.

## Description
Currently there is no way to know the ping/region of the host which makes finding games a bit hit or miss since there is no way to know if your going to have a quality connection with them.  We noticed this in testing recently I had no idea why the connection was slow until we realized where we were both playing from.

### Expected behavior
The lobby should display either the ping or the users location.  Location could either be provided by user in the settings or determined by IP.

### Actual behavior

Neither Ping or Region is displayed

### Steps to reproduce the bug

1. Open Retroarch
2. Check the Netplay Tab

### Version/Commit
You can find this information under Information/System Information

- RetroArch: [1.4.1/ba98ecd]

### Environment information

- OS: [Windows]

**Comment by RobLoach** [comment]
`connectable` flag on the server: https://github.com/RobLoach/libretro-netplay-registry/pull/24

Testing connection from the client rather then the server is a good idea too. May be better to just have it on the client-side. Thoughts?

**Comment by bparker06** [comment]
I vote not to force display of a location as the host might not want that. Also looking up a location by IP address can be wildly inaccurate for many reasons.

The room name set by the host could simply include their location if they desired (which can be done right now).

As for "ping" or latency, it's also not reliable to try a standard ICMP ping from the client to the host as it may be blocked or delayed anywhere inbetween the two. Some OSes give very low priority to ICMP packets.

My suggestion for this is to have the host support a simple netplay "ping" or heartbeat command through the TCP socket which just generates a dummy reply (like "pong") but this would be used to measure the round-trip time. The client can start a msec-level (or greater) timer when the ping request is issued. Then when the pong is received from the server, stop the timer and this would give you the latency between you and the host (not excluding any application-level latency that may need to be accounted for).

**Comment by fr500** [comment]
@RobLoach it would mean polling potentially thousands of hosts from the client for connectability. Checking if a host is connectable on first announce is good enough.

That said, I like @bparker06's idea but I don't really know how to do it, it would have to calculate a timestamp when pinging and on reply I guess and use the difference and use the reply as connectable too.
What's the best moment? when hitting refresh? or when you select the lobby? What if there are 100000 lobbies, are we gonna flood a request? Are we gonna throttle ~100ms?

It's really difficult to address these issues.
Multiplayer games refresh on demand and then continuosly on the server you're intending to connect.

**Comment by bparker06** [comment]
@fr500 Yes that approach with the timestamp works also. Perhaps only send the ping request when the room becomes visible on the screen (and optionally only refresh that again if you click refresh first?). That way if there's a ton of rooms, it will only check the latency if you've scrolled far enough to see it.

**Comment by fr500** [comment]
Then we can't achieve what's being asked (sorting by ping)

**Comment by bparker06** [comment]
The way that some other games do it is they queue off batches of several requests at once and then the list automatically re-sorts itself as the new responses come in, but I don't think that is practical for how XMB lists work.

**Comment by silentq** [comment]
You could have the host optionally provide their location in the settings.  From there organize the menu by region and then the sub menu of the region would be similar to what we have now.  I know its not the best way but at least it may avoid trying to connect to somebody super far away and cut down on high ping matches.  Again just an idea.

**Comment by RobLoach** [comment]
Don't think region or location is necessary. Just ping from the socket.

We could geocode it from the server, but I really don't think there is any benefit from it.

**Comment by LibretroAdmin** [comment]
@Cthulhu-throwaway Was this implemented? If not, do you think it'd be a good idea?
